### PR TITLE
8345375: Improve debuggability of test/jdk/java/net/Socket/CloseAvailable.java

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this test-only change which should improve the debuggability of the `test/jdk/java/net/Socket/CloseAvailable.java` test?

The test has been intermittently failing in our CI as noted in https://bugs.openjdk.org/browse/JDK-8269526. With these added logs in the test, I think it should help in the failure investigations in any future failures in this test.

Repeat testing of this test in our CI, with this change, has passed. A tier2 testing is in progress.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345375](https://bugs.openjdk.org/browse/JDK-8345375): Improve debuggability of test/jdk/java/net/Socket/CloseAvailable.java (**Sub-task** - P4)

### Reviewers
 * [Christian Stein](https://openjdk.org/census#cstein) (@sormuras - Committer)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22511/head:pull/22511` \
`$ git checkout pull/22511`

Update a local copy of the PR: \
`$ git checkout pull/22511` \
`$ git pull https://git.openjdk.org/jdk.git pull/22511/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22511`

View PR using the GUI difftool: \
`$ git pr show -t 22511`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22511.diff">https://git.openjdk.org/jdk/pull/22511.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22511#issuecomment-2514213172)
</details>

### CODE REVIEW

File: test/jdk/java/net/Socket/CloseAvailable.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
  * @test
  * @bug 4091859 8189366
  * @library /test/lib
- * @summary Test Socket.available()
+ * @summary Test Socket.getInputStream().available()
```
Comment (sormuras) [inline test/jdk/java/net/Socket/CloseAvailable.java:28]: Bit-rot in documentation, good catch.

File: test/jdk/java/net/Socket/CloseAvailable.java [hunk 2]
```diff
@@ -48,110 +48,135 @@ public static void main(String[] args) throws Exception {
...
+        is.close(); // close the inputstream and thus the underlying socket
+        System.out.println("closed inputstream of socket: " + acceptedSocket);
         try {
-            is.available();
-        }
-        catch (IOException ex) {
-            error = false;
+            final int av = is.available();
+            // available() was expected to fail but didn't
+            throw new AssertionError("Socket.getInputStream().available() was expected to fail on "
```
Comment (sormuras) [inline test/jdk/java/net/Socket/CloseAvailable.java:87]: Throwing an assertion error directly here is much better then the former fall-through approach.

File: test/jdk/java/net/Socket/CloseAvailable.java [hunk 3]
```diff
@@ -48,110 +48,135 @@ public static void main(String[] args) throws Exception {
...
-
-                InputStream is = soc.getInputStream();
-                int b = is.read();
-                assert b == 0x42;
-                assert !s.isClosed();
-                if (readUntilEOF) {
-                    b = is.read();
-                    assert b == -1;
+        final InetAddress addr = InetAddress.getLoopbackAddress();
+        try (final ServerSocket ss = new ServerSocket()) {
```
Comment (sormuras) [inline test/jdk/java/net/Socket/CloseAvailable.java:102]: Seems like now the server socket is automatically closed after usage. Good!

**Review by sormuras — APPROVED** [review]
The logic of the tests didn't change by this refactoring - the logging output printed to stdout while running and also the detail messages added to `assert`-statements will improve the debuggability.

File: test/jdk/java/net/Socket/CloseAvailable.java [hunk 4]
```diff
@@ -48,110 +48,135 @@ public static void main(String[] args) throws Exception {
...
+                            System.err.println("exception in " + Thread.currentThread().getName()
+                                    + ": " + e);
+                            e.printStackTrace();
+                        }
+                    });
+            // accept the client connect
+            acceptedSocket = ss.accept();
+            System.out.println(ss + " accepted connection " + acceptedSocket);
+        } // (intentionally) close the ServerSocket
+
```
Comment (dfuch) [inline test/jdk/java/net/Socket/CloseAvailable.java:80]: I wonder if we should join the thread here? Would that make the test more statble?

**Review by dfuch — APPROVED** [review]
Good to have more diagnosis.

File: test/jdk/java/net/Socket/CloseAvailable.java (hunk 4 above)
Comment (jaikiran) [reply test/jdk/java/net/Socket/CloseAvailable.java:80]: > probably best to leave the two close() calls (client/server) run concurrently.

I agree. Plus that will retain the existing behaviour in this test.

**Comment by jaikiran** [comment]
Thank you Christian and Daniel for the reviews. tier2 testing and repeat testing of this change went fine.

**Pull request closed without merging**